client/src/main/kotlin/io/spine/examples/pingh/client/LoginFlow.kt

```diff
@@ -0,0 +1,286 @@
+import io.spine.examples.pingh.sessions.buildBy
+import io.spine.examples.pingh.sessions.command.LogUserIn
+import io.spine.examples.pingh.sessions.command.VerifyUserLoginToGitHub
+import io.spine.examples.pingh.sessions.event.UserCodeReceived
+import io.spine.examples.pingh.sessions.event.UserIsNotLoggedIntoGitHub
+import io.spine.examples.pingh.sessions.event.UserLoggedIn
+import io.spine.net.Url
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+/**
+ * The control flow for the user login process.
+ *
+ * The flow consists of two consecutive stages. To successfully complete the login process,
+ * first enter a username and obtain a user code, then verify the login.
+ *
+ * @param client enables interaction with the Pingh server.
+ * @param session the information about the current user session.
+ */
+public class LoginFlow internal constructor(
+    private val client: DesktopClient,
+    private val session: MutableStateFlow<UserSession?>
+) {
+    /**
+     * The current stage of the login process.
+     */
+    @Suppress("MemberVisibilityCanBePrivate" /* Accessed from `desktop` module. */)
+    public val stage: MutableStateFlow<LoginStage> = MutableStateFlow(LoginStage.USERNAME_ENTERING)
+
+    /**
+     * Stores the event received after the user enters their name.
+     *
+     * This is required to initialize the verification stage flow.
+     */
+    private var userCodeReceived: UserCodeReceived? = null
+
+    /**
+     * Initiates the username entering stage of the login process.
+     *
+     * @throws IllegalStateException if the stage of the login flow is not `USERNAME_ENTERING`.
+     */
+    public fun startUsernameEntering(): EnterUsername {
+        check(stage.value == LoginStage.USERNAME_ENTERING) {
+            "Stage of the login must be `USERNAME_ENTERING`."
+        }
+        return EnterUsername(client, session) { event ->
+            stage.value = LoginStage.VERIFICATION
+            userCodeReceived = event
+        }
+    }
+
+    /**
+     * Initiates the verification stage of the login process.
+     *
+     * @throws IllegalStateException if the stage of the login flow is not `VERIFICATION`.
+     */
+    public fun startVerification(): VerifyLogin {
+        check(stage.value == LoginStage.VERIFICATION) {
+            "Stage of the login must be `VERIFICATION`."
+        }
+        return VerifyLogin(client, session, userCodeReceived!!)
+    }
+}
+
+/**
+ * The control flow of the username entering stage of the login process.
+ *
+ * @param client enables interaction with the Pingh server.
+ * @param session the information about the current user session.
+ * @param toVerificationStage called when user codes are successfully received.
+ */
+public class EnterUsername internal constructor(
```

Comment:
Let's get it clear. The previous naming approach was not good. I asked to change it to verbs. But we still have the same odd wording ("username entering") all over the place: in method names, in enums (BTW, see my comment on it below). Please do a more thorough job addressing the comments, or ask if you aren't sure.